### ISSUE TITLE
Suggestion: use lock-files instead of `Mutex`

### ISSUE THREAD

**Issue by hemisphera** [microsoft/navcontainerhelper#3285]
**Describe the issue**
I would suggest to stop using the `Mutex` class as a way to synchronize different processes accessing the same resource. The idea behind it is obvious, but I think that `Mutex` is a bad choice for the following reasons:

1. `Mutex` is not thread-safe. Or better: you have to release it on the same thread that it was created on. This of course is not a problem for the single-threaded nature of PowerShell. But for the sake of interoperability (thinking C# / .NET here) where there is a lot of asynchronous stuff happening, this becomes an issue. Example: I would like to write some .NET code that plays nicely with possible instances of BcContainerHelper running, but at the moment the I cannot synchronize my code to yours because of the thread issue of the `Mutex`.
2. `Mutex` appears to be cross-platform, but it really is not. Linux (afaik) does not have any support for system-wide named synchronization primitives. Some distros do in some way, others don't. If you run on one that doesn't, PowerShell will then just act as if there was no lock present at all. It's a niche problem, yes. But seeing you moving more and more in the Linux direction you should consider this fact.
3. `Download-Artifacts` for instance allows my to specify a different base folder to download the artifacts in. But the `Mutex` will still be the same and system-wide. So if I have two different processes downloading an artifact in two different folders, they will still lock each other - for no reason.

All of the above would be solved with a lock-file being placed as a semaphore in any specific folder. I am happy to provide a PR for it if you want to know exactly what I mean.

**Comment by freddydk** [comment]
Could you provide a small code sample of how this would work?

**Comment by hemisphera** [comment]
> Could you provide a small code sample of how this would work?

Pushed to my fork.

**Comment by freddydk** [comment]
Looks ok
The original mutex code was supplied by @jwikman (as far as I can remember) , would like to get his comments as well.

**Comment by jwikman** [comment]
Hi @hemisphera,
I understand your arguments, and if it is possible to create a 100% robust solution that works cross platform - I'm open for a change.

But I think there are a few loopholes in your implementation that could result in unwanted behavior now and then.
We added the current mutex implementation after having a lot of issues when running pipelines in parallell on the same machine, and I do not want to re-introduce those (random as it seems) errors. 😉

I've added a few comments in your code, let me know what you guys think about them. Also try to find other loopholes if you can.

When we're happy and confident with the implementation, I can apply this to our build servers to see if we can trigger any concurrency issues.

**Comment by freddydk** [comment]
On the Linux topic, linux "machines" are more atomic. Currently, I don't see multiple agents running and sharing on one Linux machine in the same way as they are running on Windows (I might be wrong)
In AL-Go for GitHub (which is one of the drivers for Linux), we are using GitHub cache to cache BC artifacts and when we (at some point in time) will have Linux containers for BC, they will be much smaller than the fairly large Windows Containers.
Add to that, Ubuntu (which is the distri behind GitHub runners, ADO hosted agents and also GitHub CodeSpaces) does actually support mutex.

**Comment by hemisphera** [comment]
> I've added a few comments in your code
@jwikman I checked them out. Yes, it could use a bit of hardening here and there, I agree. But I also don't think that some of the race conditions you describe are really relevant. For example: treating the time between checking for the existence of a file and creating it is a bit overly cautious, How probable is that really? We're not talking about millions of runs per second. If that minimal time becomes an issue, we'd have problems like this all over the place. This implementation is actually a port from a C# class I wrote and I load-tested that one without issues. But sure, even a low risk is never zero :)

> Add to that, Ubuntu (which is the distri behind GitHub runners, ADO hosted agents and also GitHub CodeSpaces) does actually support mutex.
@freddydk Ah that's good to hear. So this one at least is cleared.

Anyway guys ... At this point I'll let you decide if you want to move forward with. The (possible) benefits in the end maybe don't justify the risks?
